### DISCUSSION TITLE
Displaying IP address and MAC address on device info

### DISCUSSION THREAD

**Discussion by bdraco** [home-assistant/architecture#717] [General]
It is difficult for a user to find the ip address or mac address of a discovered device even though we know it without digging though `.storage`.  Some integrations offer a configuration url to solve this, but many devices to not have a webserver or show a cloud url which makes the configuration url only a partial solution.

Users want to lock the dhcp reservation of a device to prevent the ip from changing, but it can be hard to locate on their router if it provides a client list at all.  Home Assistant has much better discovery than most routers which means it likely a better source of the data for them.

The mac address is already stored in the device info if we know it.  To make this possible, we also need to store the last known ip address in the device info.

Proposal is to store the last know ip address in the device info under a new field as we cannot use connections because we don't want to mix the ip with stable identifiers.

Not part of this proposal, but in the future we could also use the data to improve the error when a config entry cannot setup because it is connecting to an old ip address.

**Comment by elupus** [comment]
Not fully same as this. But MAC might resemble. I lack a identification field. For example with rfxtrx, the original device id cant be seen anywhere (it's initially part of name). 

So maybe not name it as MAC, more have an hardware address or similiary named field.

**Reply by bdraco** [reply]
For connections it’s probably pretty easy to display them all, since we know what they are and how to describe them in the UI.   I’m not that familiar with rfxtrx , but is it something you can add to the connections field so we can know how to show it?

**Reply by bdraco** [reply]
But I guess it’s not really a connection.  I can’t think of a good way to display that in a normalized way for every integration. It might just be beyond the scope of this proposal.  Open to suggestions if you have an idea.

**Reply by elupus** [reply]
Somewhat hard. It's a non standardised ID string. So very specific to rfxtrx. There aight to be overlap with rflink and teldus stuff in someway, but i don't think it would be easy to get them to agree on a format.

**Reply by elupus** [reply]
For example a wall socket switch might have:  "AC 01bf47a:2" where AC represent signaltype/manufacturer, first number the group/house code, second the button/device.

**Comment by frenck** [comment]
I don't think we should aim for extending the device information with these things.
Especially since in many occurrences, these are entities as well. Thus I don't support this proposal.

As an alternative, maybe we could make it possible for the device info to point to an entity? (e.g., by unique ID) or provide a device class that can be picked up by the UI instead?

**Reply by elupus** [reply]
Why do we need to point it? Say we define some device classes for diagnostic sensors with constant info. Or a new entity category for constant info.

**Reply by frenck** [reply]
> Why do we need to point it? Say we define some device classes for diagnostic sensors with constant info. Or a new entity category for constant info.

Because else we have duplicate data. Additionally, device information doesn't update at runtime, sensors do. And device info is almost impossible to automate with, so IMHO entities are the better option.

We are now discussing a proposal here to solve a UI issue at this point IMHO.

Additionally, the mac address may already be known in the connection of the device info, and the IP may change during runtime.

**Reply by bdraco** [reply]
> Additionally, the mac address may already be known in the connection of the device info, and the IP may change during runtime.

If the ip address changes at run time, Home Assistant won't be able to connect anymore. At this point either config entry is likely going to be reloaded so the device is going to get updated with the new ip address, or the data is stale and its still useful since it will show the old ip which is the reason the device is offline in Home Assistant.

**Reply by frenck** [reply]
> If the ip address changes at run time, Home Assistant won't be able to connect anymore.

Right, but not everything uses IP address to connect? A hostname is viable and doesn't have to change.

**Reply by frenck** [reply]
In general, people like to automate, monitor and control on datapoints in Home Assistant. Via templates, markdown cards in lovelace, dunno what people come up with. Device info isn't a friendly datapoint to work with, especially since we have a great other concepts for that.

**Reply by bdraco** [reply]
For the ip it is more of a property of the config entry.  

We could add to the menu as "Show connection" if `CONF_HOST`, `CONF_IP_ADDRESS` or `CONF_URL` was present and it would display that in a popup?

**Reply by balloob** [reply]
I don't think that all device info needs to be able to be used for automations.

What we want is a system health but for devices with just info the user can act on.

**Reply by balloob** [reply]
The benefit of a system health-like approach is that we pull the data on demand, the integration doesn't need to push it all the time for when it's needed.

**Reply by frenck** [reply]
Right, this is more about the entry than the device indeed maybe? That is an interesting point.

Not sure where we should display such a thing in the entry data case.

**Reply by bdraco** [reply]
Maybe click on the title and it will pop up the unabridged title and the details (entry health could be added to that in another turn)

**Reply by frenck** [reply]
Maybe something we could ask Matthias to look into? (from an UX perspective that is)

**Comment by Jc2k** [comment]
I can't find the link on my phone but I'm sure the device registry docs used to talk about having a topology view. Would be great to show these on something like that.

**Reply by Jc2k** [reply]
(With thread it's conceivable HomeKit_controller will be using nano leaf devices as border routers for eve radiator valves so topology view could get very interesting).

**Comment by thecode** [comment]
I think that displaying MAC (maybe call it hardware address to make it compatible with other devices) and IP address is very useful, but not exactly as the suggestion here, as device info is also part of entities, I think this should just be part of device registry data but not connected to a specific entity. 
In my opinion this information is very useful for:
- Identifying a specific device (many devices has the mac printed on the device) so if you add multiple devices you can easily identify which device is which
- Reserving an IP in the router as suggested above
- Help the user detect a problem in cases the IP has changed

**Reply by johnypony3** [reply]
I just redid my network and every device has a new ip. Its all dhcp/dyns and i took an export if ma:ip, so i will get through it. But anyone who didn't do research/disco will have a very hard time.

**Comment by balloob** [comment]
I like the idea that the user can find this data. Although not the device info page itself, as it's getting crowded and this technical information is not needed 99% of the time. So instead I would suggest allow opening a dialog from the device info page with "extra/advanced/technical info".

For the IP address, I wonder if we can see if a device tracker entity is linked to the device and if so, extract the IP address from there (`ip` is a state attribute on those devices).

**Reply by thecode** [reply]
I like this idea as it allows to have a clean UI and an option to have additional space for technical information. It also makes storing this data under a section for technical data, for example we can add a key called `technical_info` stored under the device and when opening the technical info for the device the data would just populate on the card, for example:
```json
{
   "config_entries":[
      "something"
   ],
   "connections":[
      [
         "mac",
         "some-mac"
      ]
   ],
   "manufacturer":"...",
   "technical_info":[
      {
         "name":"IP Address",
         "value":"192.1.2.3"
      },
      {
         "name":"MAC Address",
         "value":"ab-cd-ef-12-34-56"
      }
   ]
}
```

**Comment by elupus** [comment]
Isn't part of this need "solved" by the new device/integration diagnostics? 

We could have a few fields standardised in that data?

**Comment by Mariusthvdb** [comment]
hope this is related, if not, ignore please:

we used to have all that info available on entity_registry_updated event, which enabled us to do: https://github.com/home-assistant/core/issues/65188

with the new device_tracker behavior, that is no longer possible for device_trackers without config entry, which are now added in a disabled state. This made it even more difficult to find the info you're after in OP. Let alone automated. because:

<img width="472" alt="Schermafbeelding 2022-02-18 om 09 46 34" src="https://user-images.githubusercontent.com/33354141/154649041-9414ade2-25d9-49af-8d5a-e8cbb4ca0d01.png">

Please consider that too when you change the logic for IP/Mac on device info

**Comment by rikroe** [comment]
I am looking at this from a different angle (although I get the discussion e.g. relating to hostname/MAC). For a connected car integration such as BMW Connected Drive I would like to be able to display the car's identification number (VIN) in the device registry (some reasons below).

It could be stored as a sensor, but that would be just an additional entity that will never ever change (like a MAC address). (Slowly) changing identification data (such as current IP address) could still be stored as sensor at the integration developer's discretion.

In BMW Connected Drive (and many other integrations I have checked), this basic identification info is already stored in the DeviceRegistry's `integrations` attribute.
Therefore I would propose to either always display the actual value of the first identifier (i.e. `this.device.identifiers[0][1]`) or add a flag to the device registry (or somewhere else) so integration developers can opt-in into this feature - it will probably not make sense everywhere.

Example screenshot:
<img width="332" alt="image" src="https://user-images.githubusercontent.com/42204099/189523453-bac13e0d-c9c8-4a9d-bd1b-d7b57aca4025.png">

Reasons for displaying this identifier in a connected car integration:
- At least for BMW, you can have multiple cars of the same type and have no opportunity to distinguish them as you cannot rename them. In the app, you at least see the image and the VIN.
- Sometimes you need your VIN for something (e.g. checking if spare parts fit). A known and easy to find place would be great in this case.

**Reply by johnypony3** [reply]
This is one of the few annoyances with ha. I got around this specific one by setting up a dhcp service and using hostnames. Takes a bit more legwork but then benefits far outweigh that.

That being said, this is not a task for the general user, even advanced ha users may not be able to set this up properly and can sour on ha altogether.

Which naturally brings up the final question. Why is it not possible to get all of the info about the integration/device? Its obviously not a complex task for the devs (given the quality/breadth of the updates).

**Reply by paulbusse** [reply]
DHCP is not always possible. It depends on the number of devices(+20), if you have control over the devices(I don't as they belong to my children, grandchildren, brothers and sisters) and whether or not they use randomized MAC addresses.
I use DHCP for the people living here permanently, but otherwise it is a coming and going of people. I trust those people as they know my WIFI password. But I have no idea about IP or (randomized) MAC address up front

**Reply by johnypony3** [reply]
im not sure why you'd be writing automation on items not stationed at your home. dhcp is giving an ip, if the they can access your network, they have an ip. i have 70+ devices on my network, each has a friendly name. and while im here 🤓, consider putting your guests on a guest network and segregate your iot from everything.

**Comment by frenck** [comment]
Fast-forward a bit, we now have a MAC address show, and support listing the visitable URL.

For the IP address... I'm not sure if we should still do that?

**Reply by bdraco** [reply]
I would still find it very helpful as I've had the displeasure of helping out on multiple issue reports where we spend hours troubleshooting a device only to find out that it's the wrong device because the user got the IP address wrong.

A visitable URL is a great solution for devices with an embedded web server, but they don't seem to be the problem devices that generate issue reports.

**Reply by joostlek** [reply]
I can't access my router so I have to check my config entry data manually to get the IP address for debugging.

**Reply by dgomes** [reply]
some devices have both Wifi and Ethernet interfaces and using the right interface makes all the difference. debugging these situations many times resorts to looking into the config entries.

**Reply by bdraco** [reply]
Thats a frequent pattern. Walking the user through getting the IP Address out of the config entry can sometimes be the difference between fixing an issue or not.  Sometimes that takes more time than actually fixing the problem.

**Reply by Jc2k** [reply]
Matter already shows the IP on the device info page, so I guess a +1 there too.

**Reply by thecode** [reply]
2 out of 3 IP based integrations I am maintaining (Switcher, Web OS TV) doesn't have a web server so there is no URL, the IP is important since it is the first thing I ask users to check if something doesn't work.

+1 from me

**Reply by mib1185** [reply]
I'm not entirely sure about showing the IP on the device overview in the UI.
When we speak about the configured (_or discovered_) ip, we can just ask the user for the diagnostics data (_if it is not implemented, than showing the IP on the device-info might also not be implemented_). Further adding the ip address in the device info box would also blow up this box and may reduce the overview on it 🤔

**Reply by balloob** [reply]
I'm in favor of a Technical info dialog that shows IP and also a lot of technical info shown today by mqtt, zwave and matter can move there (like node ID etc)

**Reply by thecode** [reply]
I like the idea of a technical info dialog (same as what is called `MQTT INFO` today, this will allow more flexible way of adding important device data without the need for each field to be specifically defined and supported by the frontend. I would even suggest to rename the what is today "MQTT INFO" to "MORE INFO" (or any other better generic name) so the UI can be similar to all devices.

This option will not clutter the device card and will allow adding information related to the device which is not necessarily common to all devices. I had a similar discussion when MAC address was added. For example Switcher has their own "Device ID" which is not the last digits of the MAC address as most devices use but it is shown on their app so if it is displayed users can easily identify the device in the app.

**Reply by brianfinley** [reply]
Yes.  MAC Address (or hardware address) and IP Address are very fundamental and important bits of information.

I found this thread while looking for info on how to contribute code to incorporate this data into the device registry so that:

1. I can correlate devices between my vendor apps (e.g.: Tapo) and HA, so that I can apply vendor app only settings and know it's going against the right HA device.
2. I can correlate devices I see on my network with devices in HA.

Here's a real-world scenario:

- I install a bazillion Tapo matter devices, using HA only and ignoring the Tapo app
- I discover that I still have to use the Tapo app to customize certain device settings
- I open the Tapo app, and it "discovers" the bazillion devices (and shows their MAC and IP, by the way), but I have no easy way to search for a device in HA by either MAC or IP

Desired resolution (I'm willing to code this if needed):
- In the registry, for each device, include an attribute for the following:
  - IP Address (include each associated IPv4 and/or IPv6 address)
  - Hardware Address (could be MAC, BLE, etc.)
- Not all devices will have all info, so it will be added opportunistically
- IP Addresses may change, and will be updated as they change
- Hardware addresses are unlikely to change, so assume they are fixed
  (is there an HA policy on how devs should treat persistence of hardware addresses?)
- I'm kinda OK with the "technical_details" structure proposed in the post above by @thecode, although I'd say that much of the info in the registry is "technical detail".  However, I'd propose that we simply use the existing "connections" attribute that supports tuples to add this data:
  - Here's a (partially redacted) example entry from my system:  
    `"connections":[["bluetooth","1B:xx:A9:xx:xx:8D"]]`
  - It already has an acceptable format (tuples)
  - It makes logical sense, in that we're talking about addresses used for communication (via some kind of connection)
  - For something like a BMW, if the VIN is used as a device identifier by an HA Integration to communicate with the vehicle service, then it would also count
  - Example:
```
    "connections":[
     ["bluetooth","AB:xx:A9:xx:xx:8D"],
     ["MAC","52:54:00:AA:BB:C1"],
     ["IP","1.1.1.2"],
     ["IPv6","2606:4700:4700::1111"],
     ["VIN","WBA8Z5C54GGS38106"],
    ]
```

**Comment by johnypony3** [comment]
> Fast-forward a bit, we now have a MAC address show, and support listing the visitable URL.
> 
> 
> 
> For the IP address... I'm not sure if we should still do that?
> 
> 
> 
> 

People have very many devices and this entity is hidden by default. Either make it easy to enable this entity for all devices or simply give the people what they are asking for and always show it.

**Reply by mib1185** [reply]
Which entity you're referring to? The mac address is not shown by any additional entity, but on the device info box in the ui:

![image](https://github.com/home-assistant/architecture/assets/35783820/e41ae7f8-0678-4cc8-b934-a14f8e7f67e4)

**Reply by johnypony3** [reply]
IP Address

**Reply by mib1185** [reply]
I guess you're referring to a specific integration which adds such entity at the moment?

**Comment by johnypony3** [comment]


**Reply by frenck** [reply]
Please... don't use email to respond to github, stuff get out of context.

**Reply by johnypony3** [reply]
Deleted and added to thread.

**Comment by drueppela** [comment]
Are there any plans to

- show the IP Address on the device info
- to have a table of all devices incl. IP Addresses and MAC Addresses

?

**Reply by johnypony3** [reply]
for two years this thread has been around and nothing

**Reply by dxmnkd316** [reply]
This is going to be dependent on individual devices.  All of mine that have IP addresses show the "Visit" button as well as the MAC address from integrations like Brother, ESPresense, ESPHome, Pi Hole, Yamaha (though some of these are HACS integrations).  Not all devices or their integrations will gather that information and display it.  If there's a specific integration that you're looking for that, I might suggest raising a feature request in the forums 

If you want a table of IP addresses and MAC addresses, I would suggest using a custom integration like Network Scanner or NMAP Tracker.  You can then setup a dashboard or card that displays them in a table:
![image](https://github.com/user-attachments/assets/95c63594-8eb7-4650-95bd-66aefec10c47)

```
## Devices
<table border="1" width="800"><tr><th>IP Address</th>
<th style="width:70%">Custom Name</th>
<th>Custom Description</th>
<th>Hostname</th>
<th>Vendor</th></tr>
{% for device in state_attr('sensor.network_scanner', 'devices') %} <tr>
  <td>{{ device.ip }}</td>
  <td>{{ device.name }}</td>
  <td>{{ device.type }}</td>
  <td>{{ device.hostname }}</td>
  <td>{{ device.vendor }}</td>
  </tr>
{% endfor %}
</tr></table>
```
